### PR TITLE
refactor!: overall revamp of the type/field code and typing improvements

### LINKED ISSUE #136: Better typing for exposed api (e.g. strawberry.django.type, strawberry.django.field, etc)
The current state of this project api is lacking typing in a lot of places.

Even though some of those use `*args`/`**kwargs` to pass not implemented parameters up to `strawberry`, static type checkers like `pyright` doesn't understand that syntax, which means that everything should be explicitly typed, with cautelous consideration in places where overloads and `__dataclass_transform__` should be sed.

This work is already done on [strabwery-django-plus](https://github.com/blb-ventures/strawberry-django-plus) and it can be ported from there. The cases that come to my mind right now are:

- `strawberry.django.type`
- `strawberry.django.input`
- `strawberry.django.partial`
- `strawberry.django.field`
- `strawberry.django.mutation`
- `strawberry.django.ordering`
- `strawberry.django.filter`

There might be other places as well

### LINKED ISSUE #137: Improve attribute resolving by avoiding `sync_to_async` overhead when the value is already cached
[strawberry-django-plus](https://github.com/blb-ventures/strawberry-django-plus) has a custom field implementation which tries to retrieve model attributes from its cache, then it fallsback to calling `sync_to_async(getattr)(model, attr)`.

This is useful because usually the attributes will be already cached and it is safe to retrieve it directly, without having the `sync_to_async` overhead.

This and other field improvements can be ported here.

### PR DESCRIPTION
**Pull request by bellini666**
This is the initial merge of a lot of improvements from
[strawberry-django-plus](https://github.com/blb-ventures/strawberry-django-plus):

* Strawberry django fields will not resolve their types lazily, just
  like strawberry core fields does. This is a major improvement for
  future annotations and other typing corner case issues.

* Django fields will now try to resolve their values from the django's
  object cache first, then fallback to the "safe resolver". This
  should remove some of the overhead `sync_to_async` causes when
  we don't actually need it.

* Since this is a major refactor, all unused and deprecated code was
  removed.

* Everything in this lib was properly typed. This means that
  `strawberry_django.type(...)`/`strawberry_django.field(...)`/etc
  will now properly validate its arguments and offer proper suggestions
  for the user.

* The code is 100% free of pyright issues (except the tests). There's
  now a new CI test that will run pyright on the codebase.

This should set the ground for the remaining work that needs
to be done for #139.

Fix #136
Fix #137

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -1,63 +1,98 @@
...
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: actions/setup-python@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
       - uses: pre-commit/action@v3.0.0
         with:
           extra_args: -a
```
Comment (patrick91) [inline .github/workflows/main.yml:28]: we don't need this anymore I think 😊 we use pre-commit.ci

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -1,63 +1,98 @@
...
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
+          cache: poetry
+      - name: Install dependencies
+        run: |
+          poetry install -E debug-toolbar
+      - name: Check for pyright errors
+        uses: jakebailey/pyright-action@v1
```
Comment (patrick91) [inline .github/workflows/main.yml:49]: nice!

**Review by patrick91 — APPROVED** [review]
I had a very quick look and it looks good to me 😊

File: .github/workflows/main.yml (hunk 1 above)
Comment (bellini666) [reply .github/workflows/main.yml:28]: Oh, that's nice to know! Will remove this then

**Pull request merged**